lib/elixir/lib/enum.ex

```diff
@@ -626,6 +626,93 @@ defmodule Enum do
     end)
   end
 
+  @doc """
+  Counts the enumerable stopping at `limit`.
+
+
+  If the enumerable implements `c:Enumerable.count/1`, this optimization is
+  not discarded. We simply call it and return the lower of the two numbers.
+  This means that `count_until/2` may not iterate over each item until count
+  is reached. To force enumeration, use `count_until/3`.
+
+  This is useful for checking certain properties of the count of an enumerable
+  without having to actually count the entire enumerable. For example, if you
+  wanted to check that the count was exactly, at least, or more than a value.
+
+  ## Examples
+
+      iex> Enum.count_until(1..20, 5)
+      5
+      iex> Enum.count_until(1..20, 50)
+      20
+      iex> Enum.count_until(1..10, 10) == 10 # At least 10
+      true
+      iex> Enum.count_until(1..11, 10 + 1) > 10 # More than 10
+      true
+      iex> Enum.count_until(1..5, 10) < 10 # Less than 10
+      true
+      iex> Enum.count_until(1..10, 10 + 1) == 10 # Exactly ten
+      true
+
+  """
+  @doc since: "1.12.0"
+  @spec count_until(t, pos_integer) :: integer
```

Comment:
Would we want to specify the return value of these functions as `non_neg_integer` too?

Reply:
Yep, good point.